File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        par = { ']' :'[' ,  ')' : '(', '}' : '{' }

        stack = []
        for c in s:
            if c in par:
                if len(stack) == 0 or stack[-1] != par[c]:
                    return False
                else:
                    stack.pop()
            else:
                stack.append(c)
        
        return len(stack) == 0

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> pr_map = {{')', '('}, {']', '['}, {'}', '{'}};
        stack<char> stack;

        for (auto p : s){
            if (pr_map.count(p))
            {
                if (!stack.empty() && stack.top() == pr_map[p])
                    stack.pop();
                else
                    return false;
            }
            else
               stack.push(p);
        }

        return stack.empty();
    }
};